### PR TITLE
Combine node name and task id into single string task id

### PR DESCRIPTION
**Pull request by imotov**
This commit changes the URL for task operations from `/_tasks/{nodeId}/{taskId}` to `/_tasks/{taskId}`, where `{taskId}` has a form of `nodeid:id`

### CODE REVIEW

**Comment by dakrone** [comment]
@imotov so what seems weird about this to me is the way it reads: `request.taskId().getId()`, do you think the `taskId()` method should be renamed? It feels like overloading the term "id" in this case

**Comment by imotov** [comment]
@dakrone I agree, do you have a suggestion for the name?

File: core/src/main/java/org/elasticsearch/tasks/TaskId.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    public TaskId(String nodeId, long id) {
+        this.nodeId = nodeId;
+        this.id = id;
+    }
+
+    public TaskId(String taskId) {
+        if (Strings.hasLength(taskId)) {
+            String[] s = Strings.split(taskId, ":");
+            if (s == null || s.length != 2) {
+                throw new IllegalArgumentException("malformed task id" + taskId);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/tasks/TaskId.java]: Missing a space here after `id`

File: core/src/main/java/org/elasticsearch/tasks/TaskId.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            }
+            this.nodeId = s[0];
+            try {
+                this.id = Long.parseLong(s[1]);
+            } catch (NumberFormatException ex) {
+                throw new IllegalArgumentException("malformed task id" + taskId, ex);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/tasks/TaskId.java]: same here

File: core/src/main/java/org/elasticsearch/tasks/TaskId.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private final String nodeId;
+    private final long id;
+
+    public TaskId(String nodeId, long id) {
+        this.nodeId = nodeId;
+        this.id = id;
+    }
+
+    public TaskId(String taskId) {
+        if (Strings.hasLength(taskId)) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/tasks/TaskId.java]: Curious, do we expect people to pass `null` for the taskId here, or can we throw on it?

File: core/src/main/java/org/elasticsearch/tasks/TaskId.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    public String getNodeId() {
+        return nodeId;
+    }
+
+    public long getId() {
+        return id;
+    }
+
+    public boolean isEmpty() {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/tasks/TaskId.java]: Not really sure what "empty" implies here, maybe add some javadoc for it?

File: core/src/main/java/org/elasticsearch/tasks/TaskId.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        return id == -1L;
+    }
+
+    @Override
+    public String toString() {
+        return nodeId + ":" + id;
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/tasks/TaskId.java]: I think we should special case the empty case, ":-1" looks kind of weird, maybe "`NONE`" or "`EMPTY`"?

File: rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json [hunk 6]
```diff
@@ -4,18 +4,18 @@
...
-        },
         "task_id": {
           "type": "number",
           "description": "Cancel the task with specified id"
         }
       },
       "params": {
+        "node_id": {
+          "type": "list",
+          "description": "A comma-separated list of node IDs or names to limit the returned information; use `_local` to return information from the node you're connecting to, leave empty to get information from all nodes"
```
Comment (dakrone) [inline rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json]: I'm worried about nodes that are named "`prod:data:1`", won't that cause errors during the task id parsing?

**Comment by dakrone** [comment]
Left some comments that are mostly minor :)

File: rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json (hunk 6 above)
Comment (imotov) [inline rest-api-spec/src/main/resources/rest-api-spec/api/tasks.cancel.json]: node_id is a parameter now, not part of the URL.

**Comment by imotov** [comment]
@dakrone pushed the changes to address your comments.

**Comment by dakrone** [comment]
LGTM

**Pull request merged**